src/onnx/parse_split.cpp

```diff
@@ -49,75 +49,139 @@ struct parse_split : op_parser<parse_split>
+            // either static shape or fixed dynamic_dimension for split axis
+            auto tuned_axis_len = input_shape.to_static(0).lens().at(tuned_axis);
+            std::vector<int64_t> vec_splits;
+            if(contains(info.attributes, "split"))
+            {
+                literal s = parser.parse_value(info.attributes.at("split"));
+                s.visit([&](auto v) { vec_splits.assign(v.begin(), v.end()); });
+            }
+            else if(args.size() == 2)
             {
-                num_outputs =
-                    parser.parse_value(info.attributes.at("num_outputs")).at<std::size_t>();
-                if(num_outputs != info.num_outputs)
+                auto s = args[1]->eval();
+                check_arg_empty(s, "PARSE_SPLIT: non-constant `split` input is not supported");
+                s.visit([&](auto v) { vec_splits.assign(v.begin(), v.end()); });
+            }
+            // no split attribute, input is equally divided
+            else
+            {
+                std::size_t num_outputs = info.num_outputs;
+                // the num_outputs attribute seems to be redundant since we already have
+                // node_info::num_outputs, but we can still perform an error check
+                if(contains(info.attributes, "num_outputs"))
+                {
+                    num_outputs =
+                        parser.parse_value(info.attributes.at("num_outputs")).at<std::size_t>();
+                    if(num_outputs != info.num_outputs)
+                    {
+                        MIGRAPHX_THROW("PARSE_SPLIT: num_outputs attribute " +
+                                       std::to_string(num_outputs) +
+                                       " doesn't match actual number of outputs " +
+                                       std::to_string(info.num_outputs) + "!");
+                    }
+                }
+                if(tuned_axis_len % num_outputs == 0)
+                {
+                    std::size_t chunk_size = tuned_axis_len / num_outputs;
+                    vec_splits.resize(num_outputs, chunk_size);
+                }
+                else
                 {
-                    MIGRAPHX_THROW("PARSE_SPLIT: num_outputs attribute " +
-                                   std::to_string(num_outputs) +
-                                   " doesn't match actual number of outputs " +
-                                   std::to_string(info.num_outputs) + "!");
+                    std::size_t chunk_size      = tuned_axis_len / num_outputs + 1;
+                    std::size_t last_chunk_size = tuned_axis_len - chunk_size * (num_outputs - 1);
+                    vec_splits.resize(num_outputs - 1, chunk_size);
+                    vec_splits.push_back(last_chunk_size);
                 }
             }
 
-            if(lens[tuned_axis] % num_outputs == 0)
+            if(std::accumulate(vec_splits.begin(), vec_splits.end(), int64_t(0)) !=
```

Comment:
Why should this check on line 163 be there after having split based on prior calculations using  `tuned_axis_len` and `num_outputs` in lines 149 onwards, please?

Reply:
It's for when the `splits` are given by the user (in the attribute).